src/future/killable.rs

```diff
@@ -0,0 +1,112 @@
+use {Future, Poll, Async};
+use task::AtomicTask;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A future which can be cancelled using a `KillHandle`.
+#[derive(Debug, Clone)]
+#[must_use = "futures do nothing unless polled"]
+pub struct Killable<T> {
+    future: T,
+    inner: Arc<KillInner>,
+}
+
+impl<T> Killable<T> where T: Future<Item = ()> {
+    /// Creates a new `Killable` future using an existing `KillRegistration`.
+    /// `KillRegistration`s can be acquired through `KillHandle::new`.
+    ///
+    /// When `kill` is called on `handle`, this future will complete
+    /// immediately. If `kill` has already been called, the future
+    /// will complete immediately without making any further progress.
+    pub fn new(future: T, reg: KillRegistration) -> Self {
+        Killable {
+            future: future,
+            inner: reg.inner,
+        }
+    }
+}
+
+/// A registration handle for a `Killable` future.
+/// Values of this type can be acquired from `KillHandle::new` and are used
+/// in calls to `Killable::new`.
+#[derive(Debug, Clone)]
+pub struct KillRegistration {
+    inner: Arc<KillInner>,
+}
+
+/// A handle to a `Killable` future.
+#[derive(Debug, Clone)]
+pub struct KillHandle {
+    inner: Arc<KillInner>,
+}
+
+impl KillHandle {
+    /// Creates a new `KillHandle` which can be used to kill a running future.
+    /// This function also returns a `KillRegistration` which is used to
+    /// pair the handle with a `Killable` future.
+    ///
+    /// This function is usually paired with a call to `Killable::new`.
+    pub fn new() -> (Self, KillRegistration) {
+        let inner = Arc::new(KillInner {
+            task: AtomicTask::new(),
+            cancel: AtomicBool::new(false),
+        });
+
+        (
+            KillHandle {
+                inner: inner.clone(),
+            },
+            KillRegistration {
+                inner: inner,
+            },
+        )
+    }
+}
+
+// Inner type storing the task to awaken and a bool indicating that it
+// should be cancelled.
+#[derive(Debug)]
+struct KillInner {
+    task: AtomicTask,
+    cancel: AtomicBool,
+}
+
+/// Creates a new `Killable` future and a `KillHandle` which can be used to stop it.
+///
+/// This function is a convenient (but less flexible) alternative to calling 
+/// `KillHandle::new` and `Killable::new` manually.
+pub fn killable<T>(future: T) -> (Killable<T>, KillHandle)
+    where T: Future<Item = ()>
+{
+    let (handle, reg) = KillHandle::new();
+    (
+        Killable::new(future, reg),
+        handle,
+    )
+}
+
+impl<T> Future for Killable<T> where T: Future<Item = ()> {
+    type Item = ();
```

Comment:
`Item` could be `Option<T::Item>` where when canceled returns `None`. A separate type to convey more semantic meaning makes sense to me as well, but would be missing basically all the useful stuff from `Option`.

Reply:
All of my use cases for this have been with futures that were side-effect only, usually being given to a `spawn` or similar. However, it seems that both you and @carllerche have other use cases in mind, so I'll change to use `Result<T::Item, Killed>`.

Reply:
I can't think of something in particular but there isn't much reason not to provide the generality

Reply:
I suppose at that point we can also provide a `Stream` impl.